### ISSUE TITLE
Defining a symbol gives a range error

### ISSUE THREAD

**Issue by skeetor** [cc65/cc65#1613]
I have the following definition:
V1: MAX_FRAMES = ($5000-$2000)/64
V2: MAX_FRAMES = (__CODE_LOAD__-SPRITE_USER_START)/SPRITE_BUFFER_LEN
V3: MAX_FRAMES = ($5000-SPRITE_USER_START)/SPRITE_BUFFER_LEN

When I use V1, my code assembles fine without any errors. These are the values which are hardcoded to test it.
Now when I use V2, which uses the same values as V1 it gives me a range error when I i.E. do something like this:

ldx #MAX_FRAMES

When I change it to V3 it also works. So it seems that the range of __CODE_LOAD__ is probably 32 bit? How can I get around this without hardcoding the values which kind of defeats the purpose.

**Comment by mrdudz** [comment]
what *exactly* is the error you get? did you import the __CODE_LOAD__ symbol?

**Comment by skeetor** [comment]
> what _exactly_ is the error you get? did you import the **CODE_LOAD** symbol?

As I said above, I get a range error which means that the symbol seems not to be a 8 bit value.

And, yes, I imported the CODE_LOAD symbol, otherwise I would have gotten an undefined symbol error instead. I also use CODE_LOAD in some other place, where the value fits, so the symbol is definitely defined.

When I do

`ldx #<__CODE_LOAD__`

it also works.

**Comment by mrdudz** [comment]
what about
```
MAX_FRAMES = ((__CODE_LOAD __-SPRITE_USER_START)/SPRITE_BUFFER_LEN) & $ffff
```

**Comment by skeetor** [comment]
> what about
> ```
> MAX_FRAMES = ((__CODE_LOAD __-SPRITE_USER_START)/SPRITE_BUFFER_LEN) & $ffff
> ```

I already tried that and it gives me the same error. I also tried with `& $ff` as well.


The project can be found here: https://github.com/skeetor/c64-src/tree/develop/C128/spreddi The relevant line is #74.

**Comment by mrdudz** [comment]
Thats weird - perhaps this is a dupe of the issue i reported: https://github.com/cc65/cc65/issues/1538

perhaps this works?
```
MAX_FRAMES = < ((__CODE_LOAD __-SPRITE_USER_START)/SPRITE_BUFFER_LEN) 
```

**Comment by greg-king5** [comment]
Versions one and three tell ca65 where the code will start.  Therefore, it can compute the formula.  It can know that the answer is eight bits wide.

But, ca65 can't know what `__CODE_LOAD__` will be; therefore, it gives the formula to ld65.  The linker calculates that formula after it has decided where things will sit.

ca65 knows that the import is 16 bits; it assumes that the entire formula will be the same.  But, you know better.  Therefore, you must help ca65 by using the `<` operator (as @mrdudz did).

**Comment by spiro-trikaliotis** [comment]
Intested: What about importing the symbol with an explicit data type specifier:
`.import __CODE_LOAD__ : WORD`

**Comment by skeetor** [comment]
> Intested: What about importing the symbol with an explicit data type specifier: `.import __CODE_LOAD__ : WORD`

This gives me an `Error: Address size specifier expected`

**Comment by skeetor** [comment]
> Thats weird - perhaps this is a dupe of the issue i reported: #1538
> 
> perhaps this works?
> 
> ```
> MAX_FRAMES = < ((__CODE_LOAD __-SPRITE_USER_START)/SPRITE_BUFFER_LEN) 
> ```

This works now. But if the addresses are wrong and the byte overflows, it wont be detected anymore causing strange unrelated errors at runtime.

**Comment by mrdudz** [comment]
Yeah, this cant be done better unless #1538 is fixed so you can use & $ffff in that expression

**Comment by skeetor** [comment]
Thanks! I hopt that this other issue is fixed any time soon. :)

**Comment by spiro-trikaliotis** [comment]
> > Intested: What about importing the symbol with an explicit data type specifier: `.import __CODE_LOAD__ : WORD`
> 
> This gives me an `Error: Address size specifier expected`

You're right, it should have been "near" or "absolute" instead of "WORD".

However, this does not work, either. I looked into the working of ca65 (which is not always easy to do).

Because there are external references, it cannot evaluate the expression. Thus, it sets the size of MAX_FRAMES to ABSOLUTE, which is 16 bit. Thus, the "LDX immediate" fails, because it cannot handle an 16 bit value.

I fear it is not easy to handle this. The trick with "& $FF" does not work, either, because it is still 16 bit. ca65 does not evaluate this expression but lets the linker do it; thus, it cannot know that the & $FF generates an 8 bit value.

Thus, @mrdudz is right, it is the same thing as with #1538. The way ca65 works at the moment does not allow this to be done differently. It would require to let ca65 special-case some operations where it can find out that some operations change the address size.

Another option could be to let ca65 assemble it (assuming that the result might be 8 bit), and then let the linker fail on it if the value happens to be bigger than 8 bit.

**Comment by mrdudz** [comment]
@skeetor please test if #1724 fixed this for you
